src/common/k8s-api/endpoints/helm-releases.api.ts

```diff
@@ -162,27 +162,30 @@ export interface HelmRelease {
   revision: string;
 }
 
-export class HelmRelease implements ItemObject {
-  constructor(data: any) {
-    Object.assign(this, data);
-    autoBind(this);
-  }
+export interface HelmRelease extends HelmReleaseDto, ItemObject {
+  getNs: () => string
+  getChart: (withVersion?: boolean) => string
+  getRevision: () => number
+  getStatus: () => string
+  getVersion: () => string
+  getUpdated: (humanize?: boolean, compact?: boolean) => string | number
+  getRepo: () => Promise<string>
+}
 
-  static create(data: any) {
-    return new HelmRelease(data);
-  }
+const toHelmRelease = (release: HelmReleaseDto) : HelmRelease => ({
```

Comment:
It’s better because now we can map it to use case specific models  

In this case we have removableReleases which have methods for selecting and removing release. 

If it’s class, we have no way to add functionality to it where needed without exposing functionality to everyone.

Reply:
I don't understand why you think that classes cannot have functionality added to them?

Reply:
Of course you could, but it’s a lot more complex than it should be. 

Why do you think that this should be a class instead of a object? And if so, how would you implement removableReleases?

Reply:
Why not a child class that takes the original as a constructor?

Reply:
Because current design of `ItemListLayout` expects `items` to extends `ItemObject`. That means that child class would need to duplicate all properties and methods of a parent. 

What does class give us here that object doesn't? Optimization?

Usage of classes allow developer to more easily attach god-class anti-pattern and does not make **composition** easy. 

Objects in the other hand makes **composition** very easy and simple. (e.g. in this case we want to compose HelmRelease, "Selectable" (_which has nothing to do with HelmReleases_) and add custom delete -method. 

**Note: "Selectable" is inlined here, because it's only used once. If we see that it's duplicated a lot of times and think that it's not wrong abstraction, then we will create function which makes any object `Selectable`.**)

In a **hypothesis** where `Selectable` is a abstraction `removableReleases` could be written:

``` typescript
const removableReleases = ({
  releases,
}: {
  releases: IAsyncComputed<HelmRelease[]>;
}): IComputedValue<RemovableHelmRelease[]> => {
  return computed(() =>
    releases.map(toSelectable).map((release) => ({
      ...release,

      delete: () => {
        // Something
      },
    }))
  );
};


// Notice how we can make any object selectable
// ----------------

interface Selectable {
  isSelected: boolean;
  toggleSelection: () => void
}

interface Item {
  someProperty: string
}

interface SelectableItem extends Item, Selectable {}

const selectableItems = ({ items }: { items: Item[] }): SelectableItem =>
  items.map(toSelectable);
```

Reply:
just curious, how would you add tests for this object that implements `HelmRelease` interface?

Reply:
Maybe I should ask a different question, why does this need to happen at all. Why not just have `removeRelease` be a dependency?

Reply:
> just curious, how would you add tests for this object that implements `HelmRelease` interface?

I don't see why you would want to test that? These should be tested on in some broader scope which uses it.

Reply:
@Nokel81 That would make `removeRelease` require parameters and thus add avoidable coupling and complexity.

Reply:
@nevalla this is a technical detail, and should probably not be unit-tested independently, but instead as part of bigger system. But for a good reason, I see no problem creating a technical unit-test for this. What is the difficulty in testing you see here?

Reply:
so we have HelmReleases API that have `listReleases` method. That method returns HelmReleases type of objects. This HelmRelease type of object has for example public `getVersion()` method that returns on some cases version from chart and on some cases empty string. Obviously this is something we want to add a test case that method is working as expected. So what is the right place to ensure the functionality of this public method?

Reply:
Well it doesn't matter at all if `HelmRelease` is object or class for that. If you really want to write that kind of black-box-test then you could do it like this:

```
describe("someTest", () => {
  let apiBaseGetMock: AsyncFnMock<() => HelmReleaseDto[]>;

  beforeEach(async () => {
    const di = getDiForUnitTesting();

    await di.runSetups();

    apiBaseGetMock = asyncFn();

    di.override(apiBaseInjectable, { get: apiBaseGetMock });
  });

  describe("when called", () => {
    let listReleasesPromise: Promise<HelmRelease[]>;

    beforeEach(() => {
      listReleasesPromise = listReleases();
    });

    it("calls API", () => {
      expect(apiBaseGetMock).toHaveBeenCalledWith({
        // Correct arguments
      });
    });

    describe("when API resolves", () => {
      beforeEach(async () => {
        const helmReleasesDtoStub = [
          {
            appVersion: "some-app-version",
            name: "some-name",
            namespace: "some-namespace",
            chart: "some-chart",
            status: "some-status",
            updated: "some-updated",
            revision: "some-revision",
          },

          {
            appVersion: "some-other-app-version",
            name: "some-other-name",
            namespace: "some-other-namespace",
            chart: "some-other-chart",
            status: "some-other-status",
            updated: "some-other-updated",
            revision: "some-other-revision",
          },
        ];

        await apiBaseGetMock.resolve(helmReleasesDtoStub);
      });

      it("knows version", async () => {
        const releases = await listReleasesPromise;

        const actual = releases.find(matches({ id: "some-id" }));

        expect(actual.getVersion()).toBe("Some special version");
      });
    });
  });
})
```

Reply:
@nevalla 
> So what is the right place to ensure the functionality of this public method?
The final right place does not exist yet, as codebase hostility prevents proper scope for unit tests, at least for some time still. We are working on this. While waiting for that, we can write technical unit tests as described by @jansav.

Also: instead of coupling unit tests systematically to technical details such as a "public method", we could do a lot better by starting to describe behavior in unit tests. But again, we are working in this.

Reply:
There's a lot to be said on this subject. Let's have a chat on your convenience. @nevalla? :)

Reply:
to me this pseudo test would make more sense here:
```
List helm releases
  when called
    it calls API
    it returns helm releases
```
and somewhere else 
```
Helm release
  knows version
  returns update as time diff
  etc
```

Or am I thinking this totally wrong?

> if you really want to write that kind of black-box-test then you could do it like this
do you mean that we should not write these kind of tests?

Reply:
but yes...this is not the case we need to resolve on this PR

Reply:
Hm. I don't see why we shouldn't? This is the only place that uses `listReleases` therefore only place that uses `static create()` and it benefits `listReleases` returning object instead of classes. We could even remove `listReleases` function.

Reply:
Hmm. We suppose we need to stop using PR-scope as excuse to not fix and touch stuff. It is simply something we cannot afford. If someone has the energy to make something better, cherishing that should be an overriding priority.